homeassistant/components/xiaomi_miio/sensor.py

```diff
@@ -145,21 +175,112 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
                         for variable in sensor_variables
                     ]
                 )
-
-    if config_entry.data[CONF_FLOW_TYPE] == CONF_DEVICE:
+    elif config_entry.data[CONF_FLOW_TYPE] == CONF_DEVICE:
         host = config_entry.data[CONF_HOST]
         token = config_entry.data[CONF_TOKEN]
-        name = config_entry.title
-        unique_id = config_entry.unique_id
-
-        _LOGGER.debug("Initializing with host %s (token %s...)", host, token[:5])
+        model = config_entry.data[CONF_MODEL]
+        device = None
+        if model in MODELS_HUMIDIFIER_MIOT:
+            device = hass.data[DOMAIN][config_entry.entry_id][KEY_DEVICE]
+            coordinator = hass.data[DOMAIN][config_entry.entry_id][KEY_COORDINATOR]
+            attributes = HUMIDIFIER_ATTRIBUTES_MIOT
+        elif model.startswith("zhimi.humidifier."):
+            device = hass.data[DOMAIN][config_entry.entry_id][KEY_DEVICE]
+            coordinator = hass.data[DOMAIN][config_entry.entry_id][KEY_COORDINATOR]
+            attributes = HUMIDIFIER_ATTRIBUTES
+        if device:
+            for attribute in attributes:
+                entities.append(
+                    XiaomiGenericSensor(
+                        f"{config_entry.title}_{attribute}",
+                        device,
+                        config_entry,
+                        f"{attribute}_{config_entry.unique_id}",
+                        attribute,
+                        hass.data[DOMAIN][config_entry.entry_id][KEY_COORDINATOR],
+                    )
+                )
+        else:
+            unique_id = config_entry.unique_id
+            name = config_entry.title
+            _LOGGER.debug("Initializing with host %s (token %s...)", host, token[:5])
 
-        device = AirQualityMonitor(host, token)
-        entities.append(XiaomiAirQualityMonitor(name, device, config_entry, unique_id))
+            device = AirQualityMonitor(host, token)
+            entities.append(
+                XiaomiAirQualityMonitor(name, device, config_entry, unique_id)
+            )
 
     async_add_entities(entities, update_before_add=True)
 
 
+class XiaomiGenericSensor(XiaomiCoordinatedMiioEntity, SensorEntity):
+    """Representation of a Xiaomi Humidifier sensor."""
+
+    def __init__(self, name, device, entry, unique_id, attribute, coordinator):
+        """Initialize the entity."""
+        super().__init__(name, device, entry, unique_id, coordinator)
+
+        self._name = name
+        self._device = device
+        self._entry = entry
+        self._unique_id = unique_id
+        self._attribute = attribute
+        self._available = None
+        self._state = None
+        self._skip_update = False
+        self.coordinator.async_add_listener(self.update)
+
+    @callback
+    def update(self):
```

Comment:
Overwrite this method instead:
https://github.com/home-assistant/core/blob/c149ecf2cc6a4efef00d537baed200b9f96bd4ed/homeassistant/helpers/update_coordinator.py#L322-L324
```suggestion
    def _handle_coordinator_update(self):
```

Reply:
Done